### PR TITLE
[SYCLomatic] Use array type for shared local memory

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -3055,6 +3057,14 @@ ParameterStream &MemVarInfo::getFuncDecl(ParameterStream &PS) {
     if (!getType()->isPointer())
       PS << " ";
     PS << "*";
+  } else if (AccMode == PointerToArray) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp:3060]: Move ti before if (isShared())

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 2]
```diff
@@ -3085,6 +3095,16 @@ ParameterStream &MemVarInfo::getKernelArg(ParameterStream &PS) {
         PS << "template ";
       PS << "get_multi_ptr<" << MapNames::getClNamespace()
          << "access::decorated::no>().get()";
+    } else if (isShared() && AccMode == PointerToArray) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]:
```suggestion
    } else if (AccMode == PointerToArray) {
```

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 3]
```diff
@@ -3085,6 +3095,16 @@ ParameterStream &MemVarInfo::getKernelArg(ParameterStream &PS) {
...
+      std::string CastType = getType()->getBaseName();
+      auto Range = getType()->getRange();
+      CastType = CastType + " (*)";
+      for (size_t i = 1; i < Range.size(); i++) {
+        CastType = CastType + "[" + Range[i].getSize() + "]";
+      }
+      PS << "reinterpret_cast<" << CastType << ">(" << getAccessorName()
```
Comment (zhiweij1) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Use isWritten() to check, if false, we can use static_cast, otherwise, no cast needed.

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by ziranzha — APPROVED** [review]

**Pull request merged**